tests/binning_test.py

```diff
@@ -0,0 +1,437 @@
+    table = sc.data.table_xyz(100)
+    table.coords['label'] = (table.coords['x'] * 10).to(dtype=dtype)
+    coord = table.bin(label=sc.scalar(3.3, unit='m')).coords['label']
+    expected = sc.array(dims=['label'], unit='m', dtype=dtype, values=[0, 3, 6, 9, 12])
+    assert sc.identical(coord, expected)
+
+
+@pytest.mark.parametrize('dtype', ['int32', 'int64'])
+def test_bin_integer_coord_by_fractional_stepsize_raises(dtype):
+    table = sc.data.table_xyz(100)
+    table.coords['label'] = (table.coords['x'] * 10).to(dtype=dtype)
+    with pytest.raises(ZeroDivisionError):
+        table.bin(label=sc.scalar(0.5, unit='m')).coords['label']
+
+
+def test_group_after_bin_considers_event_value():
+    table = sc.data.table_xyz(100)
+    table.coords['label'] = (table.coords['x'] * 10).to(dtype='int64')
+    da = table.bin(label=2).group('label')
+    assert da.sizes == {'label': 10}
+
+
+def test_bin_erases_dims_automatically_if_labels_for_same_dim():
+    table = sc.data.table_xyz(100)
+    table.coords['x2'] = table.coords['x'] * 2
+    da = table.bin(x=10)
+    assert da.bin(x2=100).dims == ('x', 'x2')
+    # With a bin-coord for x2 we can deduce that the new binning is by different
+    # 'labels' for the same data dimension.
+    da.coords['x2'] = da.coords['x'] * 2
+    assert da.bin(x2=100).dims == ('x2', )
+
+
+def test_bin_by_two_coords_depending_on_same_dim():
+    table = sc.data.table_xyz(100)
+    da = table.bin(x=10, y=12)
+    assert da.dims == ('x', 'y')
+
+
+def test_bin_by_2d_erases_2_input_dims():
+    table = sc.data.table_xyz(100)
+    table.coords['xy'] = table.coords['x'] + table.coords['y']
+    da = table.bin(x=10, y=12)
+    # Note: These are bin edges, but `bin` still works
+    da.coords['xy'] = da.coords['x'][1:] + da.coords['y']
+    assert da.bin(xy=20).dims == ('xy', )
+
+
+def test_bin_by_2d_dimension_coord_does_not_erase_extra_dim():
+    table = sc.data.table_xyz(100)
+    table.coords['xy'] = table.coords['x'] + table.coords['y']
+    da = table.bin(x=10, y=12)
+    da.coords['xy'] = da.coords['x'][1:] + da.coords['y']
+    da = da.rename_dims({'y': 'xy'})
+    # The call to `bin` here "aligns" all bins, so we generally do not want ot erase.
+    assert da.bin(xy=20).dims == ('x', 'xy')
+
+
+rng = default_rng(seed=1234)
+
+
+def date_month_day_table_grouped_by_date():
+    size = 100
+    table = sc.DataArray(sc.ones(dims=['row'], shape=[size]))
+    table.coords['date'] = sc.array(dims=['row'], values=rng.integers(365, size=size))
+    table.coords['month'] = table.coords['date'] // 30
+    table.coords['day'] = table.coords['date'] % 30
+    # Using bin instead of group to preserve the original `date` coord in the buffer,
+    # as needed by some tests.
+    da = table.bin(date=365)
+    da.coords['date'] = da.coords['date'][:-1].copy()
+    return da
+
+
+# In the following we test the automatic erasure of dimensions. It should be consistent
+# across the different operations, bin, group, hist, and nanhist.
```

Comment:
@nvaytet @jl-wynen See below for the new tests documenting the "erase" behavior.